### PR TITLE
first pass at #11 (domain tags)

### PR DESCRIPTION
**Pull request by briandilley**
Closes: #11

## Description

I've added support for domain tags, but I can't get any of the tests to fail when I set the domain tag to something that it shouldn't be (ie: something other than values `FLOW-V0.0-transaction` and `FLOW-V0.0-user`).

______

For contributor use:

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [standards mentioned here](https://github.com/onflow/flow-jvm-sdk/blob/master/CONTRIBUTING.md#styleguides).
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: src/main/kotlin/org/onflow/sdk/domain-tags.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    val TRANSACTION_DOMAIN_TAG = normalize("FLOW-V0.0-transaction")
+
+    @JvmStatic
+    val USER_DOMAIN_TAG = normalize("FLOW-V0.0-user")
+
+    @JvmStatic
+    fun normalize(tag: String?): ByteArray {
+        val normalizedTag = when {
+            tag == null -> null
+            tag.length > 32 -> throw IllegalArgumentException("Domain tags cannot be longer than 32 characters long")
```
Comment (psiemens) [inline src/main/kotlin/org/onflow/sdk/domain-tags.kt:15]:
```suggestion
            tag.length > 32 -> throw IllegalArgumentException("Domain tags cannot be longer than 32 characters")
```

File: src/main/kotlin/org/onflow/sdk/transaction-dsl.kt [hunk 2]
```diff
@@ -333,7 +334,8 @@ class PendingSignature(
     val address: FlowAddress? = null,
     val keyIndex: Int? = null,
     val signer: Signer? = null,
-    val signature: FlowSignature? = null
+    val domainTag: ByteArray? = null,
```
Comment (psiemens) [inline src/main/kotlin/org/onflow/sdk/transaction-dsl.kt:337]: `PendingSignature` is only ever used for transactions, right? (given that it includes `FlowTransactionSignature`). If that's the case, I would remove the `domainTag` field and just make it a static member that is always set to `DomainTag.TRANSACTION_DOMAIN_TAG`.

File: src/test/kotlin/org/onflow/sdk/ScriptTest.kt [hunk 3]
```diff
@@ -111,4 +115,108 @@ class ScriptTest {
         assertEquals(HashAlgorithm.SHA3_256, struct.hashAlgorithm)
         assertTrue(struct.isValid)
     }
+
+    @Test
+    fun `Test domain tags`() {
```
Comment (psiemens) [inline src/test/kotlin/org/onflow/sdk/ScriptTest.kt:120]: Great test!

File: src/main/kotlin/org/onflow/sdk/transaction-dsl.kt [hunk 4]
```diff
@@ -446,6 +451,18 @@ class FlowTransactionSignatureBuilder {
...
+        get() { return _domainTag }
+        set(value) { _domainTag = value }
+
+    fun domainTag(domainTag: String) {
+        this.domainTag = DomainTag.normalize(domainTag)
+    }
+    fun domainTag(domainTag: ByteArray) {
+        this.domainTag = domainTag
+    }
+    fun domainTag(domainTag: () -> ByteArray) = this.domainTag(domainTag())
```
Comment (psiemens) [inline src/main/kotlin/org/onflow/sdk/transaction-dsl.kt:464]: Same here -- the domain tag for a transaction will always be `DomainTag.TRANSACTION_DOMAIN_TAG`, so I don't think this needs to be settable.

**Review by psiemens — COMMENTED** [review]
Looks great overall, thanks for implementing this!

@briandilley Sorry, I should have mentioned this before -- the reason it works on the emulator with or without the domain tag is because the latest emulator release is backward compatible to support both. So that's expected behaviour.

File: src/main/kotlin/org/onflow/sdk/transaction-dsl.kt (hunk 2 above)
Comment (briandilley) [reply src/main/kotlin/org/onflow/sdk/transaction-dsl.kt:337]: makes sense, i thought about that myself.

File: src/main/kotlin/org/onflow/sdk/transaction-dsl.kt (hunk 4 above)
Comment (briandilley) [reply src/main/kotlin/org/onflow/sdk/transaction-dsl.kt:464]: 👍

File: src/test/kotlin/org/onflow/sdk/ScriptTest.kt (hunk 3 above)
Comment (briandilley) [reply src/test/kotlin/org/onflow/sdk/ScriptTest.kt:120]: also - it's just a port of the flow test ;)

**Pull request merged**